src/e3/anod/qualifiers_manager.py

```diff
@@ -1,40 +1,210 @@
+        )
+        self.repr_omit_key = repr_omit_key
+
+        # Check that default value is valid
+        if default is not None and choices is not None and default not in choices:
+            choices_str = ", ".join((f"'{choice}'" for choice in choices))
+            raise AnodError(
+                f"{self.origin}: default value '{default}' "
+                f"should be in ({choices_str})."
+            )
+
+        self.choices = choices
+        self._default = default
+
+    @property
+    def default(self) -> str | bool | None:
+        """See QualifierDeclaration.default."""
+        return self._default
+
+    def value(self, value: str | bool) -> str | bool:
+        """See QualifierDeclaration.value."""
+        if self.choices is not None and value not in self.choices:
+            choices_str = ", ".join((f"'{choice}'" for choice in self.choices))
+            raise AnodError(
+                f"{self.origin}: Invalid value for qualifier {self.name}: '{value}' "
+                f"not in ({choices_str})"
+            )
+        return value
+
+    def repr(self, value: str | bool, hash_pool: list[str] | None) -> str:
+        """See QualifierDeclaration.repr."""
+        list_repr = []
+        if not self.repr_omit_key:
+            list_repr.append(self.repr_name)
+        if value:
+            list_repr.append(str(value))
+
+        str_repr = "-".join(list_repr)
 
-    class QualifierTagDeclaration(QualifierDeclaration):
-        type: Literal["tag"]
+        if hash_pool is not None and self.repr_in_hash:
+            if str_repr:
+                hash_pool.append(str_repr)
+            return ""
+        else:
+            if (
+                value != ""
+                and value == self.default
+                and self.choices is not None
+                and "" not in self.choices
+            ):
+                # In the case the value of qualifier is a finite set and
+                # that "" is not in that set, if value is the default value then
+                # just return an empty representation.
+                return ""
+            else:
+                return str_repr
 
-    class QualifierKeyValueDeclaration(QualifierDeclaration):
-        type: Literal["key-value"]
 
-        # Fields specific to key-value qualifiers
-        default: str | None
-        choices: list[str] | None
-        repr_omit_key: bool
+class TagDeclaration(QualifierDeclaration):
+    """Tag qualifier declaration."""
+
+    @property
+    def default(self) -> bool:
+        """See QualifierDeclaration.value."""
+        return False
+
+    def value(self, value: str | bool) -> bool:
+        """See QualifierDeclaration.value."""
+        if value is False:
+            return False
+        else:
+            return True
+
+    def repr(self, value: str | bool, hash_pool: list[str] | None) -> str:
+        """See QualifierDeclaration.repr."""
+        str_repr = self.repr_name if self.repr_name else ""
```

Comment:
Is it possible that a qualifier don't have a representation?

Reply:
Right. Code adjusted

Reply:
Solved